### ISSUE TITLE
FaceDetection+CustomTraining+Validation +Performance validation 

### ISSUE THREAD

**Issue by abhigoku10** [natanielruiz/deep-head-pose#22]
@natanielruiz  Hi thanks for the awesome work and sharing just had few queries 

1. For face detection should we always use the FRCNN or is there any possibility of using some other detection technique
2. How to validate the output params like the pitch,roll and yaw values obtained from the algorithm 
3. Are the values of pitch,roll and yaw being generated based on the center of the face detection 
4. Can we use ur algorithm for training it on custom dataset or the model share can be used on generic dataset 
5. For custom training can you provide some references steps to achieve this 
6. I ran your code on the CPU and its very slow is there any provision in the future for running it on the CPU 

Thanks for the awesome work and sharing

**Comment by natanielruiz** [comment]
1. You can use whatever you like for detection. This code usually uses dockerface (from my github account) or I think I also put an example using dlib's detection.
2. Validate in which sense?
3. They are generated based on the cropped face image which is passed into the pose-predicting neural network.
4. You can train my model on a custom dataset. The model shared should be able to generalize to a bunch of applications, that's the main contribution of my work. But it might fail in a bunch of other cases too!
5. I think the code explains it pretty well so I suggest exploring the code and documentation.
6. Running on CPU will always be slow with networks of this size. There are no plans to deploy smaller faster networks.

Thank you, cheers! In general I suggest you read all of the documentation again at this point if you have any doubts.

**Comment by abhigoku10** [comment]
1. as far as i am aware of we can use any detection algorithm we want a bounding box file created similar to dockerface 
2.I meant to say what ever the values generated from an algorithm , how can we say its the right value obtained or not . Basically the performance of the algorithm (accuracy and timing)
3. so since you mentioned it works on the cropped face is there any alignment of the cropped which we obtain in the dlib and process it to the algorithm 
4.i am currently testing on the different cases but it a great work, work for most of my scenarios
5. i could find the documentation besides the readme file can you pls suggest the location 
6. the intention was to know can this be ported on board devices so as to work in that too 

thanks for the response

**Comment by natanielruiz** [comment]
1. I assume this question has been answered.
2. You need to have ground truth annotations of face pose to evaluate the performance.
3. No alignment from cropped face image. I'm glad it's working on most of your scenarios, I'd love to see some examples!
4. You're right, training is not documented. Please check out the training script and go from there, it's not very complicated.
5. I don't think this can be ported to work real time on anything slower than a GPU. You would have to do some engineering work for that: smaller model for example.

**Comment by exnx** [comment]
@abhigoku10 Hi!  I was curious if you knew how fast (frames per second) you were able to get this to run on your CPU?  I was going to try it on my Mac too!  Just the pretrained models.  Don't suppose you'd be interested in sharing your repo for the CPU version?  No worries if not.  Thanks!

Thanks @natanielruiz!

**Comment by natanielruiz** [comment]
For CPU just take away every .gpu() from the code @exnx

**Comment by abhigoku10** [comment]
@exnx The cpu version is a bit slow . i did not estimate the fps . I shall share the repo but still have to do a lot modifications so once done shall share , if you want i can give the pointers of the changes

**Comment by exnx** [comment]
@abhigoku10 awesome, thanks, would love to see it when you're ready.  In the mean time, was it just the removal of .gpu() to get it to run on cpu, or some other things too?  Did you use dockerface for the face detection too?  I wonder if that makes it slower for CPU as well, and perhaps a dlib face detector could be sufficient, even if you lose some accuracy, the speed gain would be worth it in use case.

@natanielruiz thanks for the tip!

**Comment by natanielruiz** [comment]
You don't lose accuracy with dlib but the center of the face is jittery. @exnx

**Comment by abhigoku10** [comment]
@exnx I did not test with dockerface since dlib was working well for me , on the timing part i need to test it but there are few optimizations required to achieve good timing .Please let me know if you have done any optimizations 
On the cpu section i have made this following changes 
1.idx_tensor = torch.FloatTensor(idx_tensor)
2.    cudnn.enabled = False 
So u can make these following changes and let me know if your facing some issues
